Data/Primitive/PrimArray.hs

```diff
@@ -0,0 +1,472 @@
+import GHC.Float (Float(..), Double(..))
+import GHC.Word (Word(..), Word8(..), Word16(..), Word32(..), Word64(..))
+import GHC.Int (Int8(..), Int16(..), Int32(..), Int64(..))
+import GHC.Ptr (Ptr(..), FunPtr(..))
+
+#if !(MIN_VERSION_base(4,8,0))
+import Data.Monoid (mappend)
+#endif
+
+-- | Primitive array tagged with element type @a@.
+--
+newtype PrimArray a = PrimArray ByteArray
+    deriving (Typeable, Data)
+
+instance (Prim a, Eq a) => Eq (PrimArray a) where
+    {-# INLINE (==) #-}
+    (==) = eqPrimArray
+
+eqPrimArray :: (Prim a, Eq a) => PrimArray a -> PrimArray a -> Bool
+{-# INLINE [1] eqPrimArray #-}
+eqPrimArray paA paB
+    | paA `samePrimArray` paB = True
+    | otherwise =
+        let !sizA = sizeofPrimArray paA
+            !sizB = sizeofPrimArray paB
+        in sizA == sizB && go 0 sizA
+  where
+    go !i !siz | i >= siz  = True
+               | otherwise = indexPrimArray paA i == indexPrimArray paB i && go (i+1) siz
+
+-- | Byte wise comparison using @memcmp@.
+--
+-- This function can be used to speed up 'PrimArray' 's 'Eq' instance by comparing array
+-- using memcmp, all built-in 'Prim' types enabled optimization already, and if your
+-- custom 'Prim' type can be compared byte wise, you can use rewrite rules to override
+-- default 'Eq' instance which use looping by default. e.g.
+--
+-- @\{\-\# RULES "eqPrimArray/Foo"   eqPrimArray = memcmpPrimArray (0 :: Foo) \#\-\}@
+--
+memcmpPrimArray :: a -> PrimArray a -> PrimArray a -> Bool
+{-# INLINE memcmpPrimArray #-}
+memcmpPrimArray _ (PrimArray baA) (PrimArray baB) = baA == baB
+
+{-# RULES "eqPrimArray/Word"   eqPrimArray = memcmpPrimArray (undefined :: Word) #-}
+{-# RULES "eqPrimArray/Int"    eqPrimArray = memcmpPrimArray (undefined :: Int) #-}
+{-# RULES "eqPrimArray/Float"  eqPrimArray = memcmpPrimArray (undefined :: Float) #-}
+{-# RULES "eqPrimArray/Double" eqPrimArray = memcmpPrimArray (undefined :: Double) #-}
+{-# RULES "eqPrimArray/Char"   eqPrimArray = memcmpPrimArray (undefined :: Char) #-}
+
+{-# RULES "eqPrimArray/Word8"  eqPrimArray = memcmpPrimArray (undefined :: Word8) #-}
+{-# RULES "eqPrimArray/Word16" eqPrimArray = memcmpPrimArray (undefined :: Word16) #-}
+{-# RULES "eqPrimArray/Word32" eqPrimArray = memcmpPrimArray (undefined :: Word32) #-}
+{-# RULES "eqPrimArray/Word64" eqPrimArray = memcmpPrimArray (undefined :: Word64) #-}
+{-# RULES "eqPrimArray/Int8"   eqPrimArray = memcmpPrimArray (undefined :: Int8) #-}
+{-# RULES "eqPrimArray/Int16"  eqPrimArray = memcmpPrimArray (undefined :: Int16) #-}
+{-# RULES "eqPrimArray/Int32"  eqPrimArray = memcmpPrimArray (undefined :: Int32) #-}
+{-# RULES "eqPrimArray/Int64"  eqPrimArray = memcmpPrimArray (undefined :: Int64) #-}
+
+{-# RULES "eqPrimArray/Addr"   eqPrimArray = memcmpPrimArray (undefined :: Addr) #-}
+{-# RULES "eqPrimArray/Ptr"    eqPrimArray = memcmpPrimArray (undefined :: Ptr a) #-}
+{-# RULES "eqPrimArray/FunPtr" eqPrimArray = memcmpPrimArray (undefined :: FunPtr a) #-}
```

Comment:
Ugh... Wouldn't it be better to use a class for `eqPrimArray`? Rules are fragile, and require user involvement, and so on. It's not great. Separately, passing `undefined` is rather old-fashioned. Use a proxy type instead.